fat/fat2/addressbalancedeltamap.go

```diff
@@ -0,0 +1,78 @@
+package fat2
+
+import (
+	"encoding/json"
+	"fmt"
+	"github.com/Factom-Asset-Tokens/factom"
+)
+
+type AddressBalanceDeltaMap map[factom.FAAddress]map[string]int64
+
+// MarshalJSON marshals the map of addresses and their balance deltas for a list of pegged assets
+func (m AddressBalanceDeltaMap) MarshalJSON() ([]byte, error) {
+	strBalanceDeltaMap := make(map[string]map[string]int64)
+	for address, deltas := range m {
+		strBalanceDeltaMap[address.String()] = deltas
+	}
+	return json.Marshal(strBalanceDeltaMap)
+}
+
+// UnmarshalJSON unmarshals a map of addresses and their balance deltas for a list of pegged assets
+func (m *AddressBalanceDeltaMap) UnmarshalJSON(data []byte) error {
+	var strBalanceDeltas map[string]map[string]int64
+	if err := json.Unmarshal(data, &strBalanceDeltas); err != nil {
+		return err
+	}
+	if len(strBalanceDeltas) == 0 {
+		return fmt.Errorf("%T: empty", m)
+	}
+	*m = make(AddressBalanceDeltaMap, len(strBalanceDeltas))
+	var adr factom.FAAddress
+
+	// TODO: replace this list with an imported const from a pegnet module
```

Comment:
@AdamSLevy Attempting to follow your suggestion, I implemented the type here: https://github.com/sambarnes/fatd/blob/fat2/fat/fat2/peggedassettickers.go

Not sure how i feel about it yet:
- there are two parallel switch statements with duplicate information
- the fact that quotes are managed manually feels hacky (just a gut reaction)

Also, it forced me introduce a change in `AddressBalanceDeltaMap` where I now have to [manually unmarshal the `PeggedAssetTicker` components of the map](https://github.com/sambarnes/fatd/blob/fat2/fat/fat2/addressbalancedeltamap.go#L38). It would always fail if I instead had `var strBalanceDeltas map[string]map[PeggedAssetTicker]int64`, is this just an inherent problem with unmarshalling nested maps with custom types?

Curious to hear your thoughts on the above and if this feels cleaner to you.

Reply:
To reduce the parallel switch statements it may be best to use a single global unexported map, but at least the logic will be contained to the type.

Unfortunately the json package can't unmarshal map keys that aren't a basic type. I find this limiting and I hope one day they correct this. So you are right that you have to first unmarshal it into a map[string]...

Fortunately this complexity is centrally located in the marshaling code for the specific type. From that point on, we can check for valid ticker values by using bounds checking on the PeggedAssetTicker (int). You might add a constant called PeggedAssetTickerMax for such comparisons. Personally I would like to see a shorter type name and I would like that all of the constants start with the type name. The rationale for starting all of those constants with the type name is that it helps with variable completion in editors and makes the constants very easy to lookup from go doc.

Reply:
Switched to an unexported map and i do like it better now. Does force the Marshal function to do a linear search through the map to find the string representation, but N is low and Unmarshals will greatly outnumber Marshals so its no big deal.

Reply:
You could always use an array for one direction and a map for the other down the road if this ever becomes a choke point.